### PR TITLE
feature: Compactable

### PR DESCRIPTION
**Pull request by raveclassic**
First part of https://github.com/gcanti/fp-ts/pull/485
Adds `Compactable` typeclass, default imeplementations of `compact` and `separate` and tests

### CODE REVIEW

File: src/Compactable.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  readonly separate: <A, B>(fa: Type3<F, U, L, Either<A, B>>) => Separated<Type3<F, U, L, A>, Type3<F, U, L, B>>
+}
+
+/**
+ * Gets default implementation of {@link Compactable.compact} using {@link Compactable.separate}
+ * @function
+ * @since 1.7.0
+ */
+export function getCompactFromSeparate<F extends URIS3, U, L>(
+  F: Functor3C<F, U, L> & Pick<Compactable3C<F, U, L>, 'separate'>
```
[the remark this answers is not in the document]
Comment (raveclassic) [reply src/Compactable.ts]: The trick is to accept an object which is both a Functor and partial Compatible with only `separate` defined. `Compactable3C<F, U, L>['separate']` returns the type of `separate`, not an object

File: src/Compactable.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import { fromEither, Option } from './Option'
+import { Either, fromOption } from './Either'
+import { Functor, Functor1, Functor2, Functor2C, Functor3, Functor3C } from './Functor'
+
+/**
+ * `Separated` type which holds `left` and `right` parts
+ * @since 1.7.0
+ */
+export type Separated<A, B> = {
+  left: A
```
[the remark this answers is not in the document]
Comment (raveclassic) [reply src/Compactable.ts]: done

File: src/Compactable.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+/**
+ * @see Compactable
+ * @since 1.7.0
+ */
+export interface Compactable1<F extends URIS> {
+  readonly URI: F
+  readonly compact: <A>(fa: Type<F, Option<A>>) => Type<F, A>
+  readonly separate: <A, B>(fa: Type<F, Either<A, B>>) => Separated<Type<F, A>, HKT<F, B>>
```
[the remark this answers is not in the document]
Comment (raveclassic) [reply src/Compactable.ts]: done

File: src/Compactable.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+export function getCompactFromSeparate<F>(F: Functor<F> & Pick<Compactable<F>, 'separate'>): Compactable<F>['compact'] {
+  return foa => F.separate(F.map(foa, fromOption(null))).right
+}
+
+/**
+ * Gets default implementation of {@link Compactable.separate} using {@link Compactable.compact}
+ * @function
+ * @since 1.7.0
+ */
+export function getSeparateFromCompact<F extends URIS3, U, L>(
```
[the remark this answers is not in the document]
Comment (raveclassic) [reply src/Compactable.ts]: I picked that names to be similar to `getMonoid`, `getSemigroup` etc. But let's rename, I'm ok with it

File: src/Compactable.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gcanti) [reply src/Compactable.ts]: I'm dumb, obviously we need an object there, nice trick

**Pull request merged**